lib/TunnelCluster.js

```diff
@@ -1,133 +1,150 @@
-var TunnelCluster = function(opt) {
-    if (!(this instanceof TunnelCluster)) {
-        return new TunnelCluster(opt);
-    }
-
-    var self = this;
-    self._opt = opt;
-
-    EventEmitter.call(self);
-};
-
-TunnelCluster.prototype.__proto__ = EventEmitter.prototype;
-
-// establish a new tunnel
-TunnelCluster.prototype.open = function() {
-    var self = this;
-
-    var opt = self._opt || {};
-
-    var remote_host = opt.remote_host;
-    var remote_port = opt.remote_port;
-
-    var local_host = opt.local_host || 'localhost';
-    var local_port = opt.local_port;
-
-    debug('establishing tunnel %s:%s <> %s:%s', local_host, local_port, remote_host, remote_port);
+module.exports = class TunnelCluster extends EventEmitter {
+  constructor(opts = {}) {
+    super(opts);
+    this.opts = opts;
+  }
+
+  open() {
+    const opt = this.opts;
+
+    // Preference IP if returned by the server
+    const remoteHostOrIp = opt.remote_ip || opt.remote_host;
+    const remotePort = opt.remote_port;
+    const localHost = opt.local_host || 'localhost';
+    const localPort = opt.local_port;
+    const localProtocol = opt.local_https ? 'https' : 'http';
+
+    debug(
+      'establishing tunnel %s://%s:%s <> %s:%s',
+      localProtocol,
+      localHost,
+      localPort,
+      remoteHostOrIp,
+      remotePort
+    );
 
     // connection to localtunnel server
-    var remote = net.connect({
-        host: remote_host,
-        port: remote_port
+    const remote = net.connect({
+      host: remoteHostOrIp,
+      port: remotePort,
     });
 
     remote.setKeepAlive(true);
 
-    remote.on('error', function(err) {
-        // emit connection refused errors immediately, because they
-        // indicate that the tunnel can't be established.
-        if (err.code === 'ECONNREFUSED') {
-            self.emit('error', new Error('connection refused: ' + remote_host + ':' + remote_port + ' (check your firewall settings)'));
-        }
-
-        remote.end();
+    remote.on('error', err => {
+      debug('got remote connection error', err.message);
+
+      // emit connection refused errors immediately, because they
+      // indicate that the tunnel can't be established.
+      if (err.code === 'ECONNREFUSED') {
+        this.emit(
+          'error',
+          new Error(
+            `connection refused: ${remoteHostOrIp}:${remotePort} (check your firewall settings)`
+          )
+        );
+      }
+
+      remote.end();
     });
 
-    function conn_local() {
-        if (remote.destroyed) {
-            debug('remote destroyed');
-            self.emit('dead');
-            return;
+    const connLocal = () => {
+      if (remote.destroyed) {
+        debug('remote destroyed');
+        this.emit('dead');
+        return;
+      }
+
+      debug('connecting locally to %s://%s:%d', localProtocol, localHost, localPort);
+      remote.pause();
+
+      const allowInvalidCert = process.env.NODE_TLS_REJECT_UNAUTHORIZED === '0';
```

Comment:
This should be set via an option rather than rely on the environment variable which affects all TLS requests node will make.

Reply:
I guess we want to accept both the env var and an option. I'll add it.

Reply:
Done.